### ISSUE TITLE
Unreliable segments flagging

### ISSUE THREAD

**Issue by DanielHoffmannAyala** [translationalneuromodeling/tapas#11]
Hello,

in your 2017 paper _"The PhysIO Toolbox for Modeling Physiological Noise in fMRI Data"_ at the end of section 2.7.3 you write:

> If unreliable segments have been detected earlier (see section Assessment of Recording Quality and Recovery), the regressors will be split and the common multiple regressors file will contain zeros at these time points. The regressors derived in the unreliable segments are saved in a separate file.

While processing our physiological data, we do find indicators in the output figures that there indeed are constant, hence unreliable data snippets in the data. Yet, these segments are neither "flagged" in the common multiple regressors (text) file (by setting their values to 0), nor are there any separate (text) files saved additionally in our save directory.
The respective Matlab variables _physio.ons_secs.c_is_reliable_ and _physio.ons_secs.r_is_reliable_ are created and properly reflect the "unreliable" data snippets' indices. But they don't seem to be used for any purpose related to modification of output regressors?
Could you elucidate this discrepancy?

If this feature is indeed not implemented: would you recommend for us to manually "censor" these loci in _physio.models.R_ by substituting the affected sections' values with 0?

Thank you very much for your help!

Regards,
Daniel

**Comment by mrikasper** [comment]
Dear Daniel,

my apologies for the conflicting information about this censoring. You are right, the current release does not include the actual creation of the regressors, only the detection of the unreliable raw data snippets.

To properly support and test this feature, we would need some better test data that has some unreliable data snippets (moving subject, belt detachment). If you have such data and are happy to share it in an anonymized way, we could implement the censored regressor creation for the next release.

What you suggest in manipulating the design matrix `R` is exactly the right way to go: Set the values to zero for the parts of the time series that were flagged. You will have to resample the flagged time series part `physio.ons_secs.c_is_reliable` to contain only the time points used in the design matrix (one per volume, time point of the `onset_slice`). You can re-use the resampling code used in, for example, `tapas_physio_create_retroicor_regressors`.

For the cardiac flagging, and the RETROICOR model this would be:

```matlab
is_reliable = ons_secs.c_is_reliable;
rsampint    = ons_secs.t(2)-ons_secs.t(1);
sample_points   = tapas_physio_get_sample_points(ons_secs, sqpar);
is_reliable_at_sample_points  = tapas_physio_downsample_phase(ons_secs.spulse, ...
 is_reliable, sample_points, rsampint);

% regressor parts related to cardiac time series
cardiac_sess(~is_reliable_at_sample_points,:) = 0;
mult_sess(~is_reliable_at_sample_points,:) = 0;

```

You could just include this in the `tapas_physio_main_create_regressors` function or write an own function `tapas_physio_censor_unreliable_regressor_parts`, which could be called just after `tapas_physio_create_retroicor_regressors`.

For RETROICOR, this is all straightforward and could be implemented that way, however, for the convolution models (HRV, RVT), it is not quite clear to me whether censoring the instantaneous impact of bad recordings is enough, since the convolution kernel will also compromise data up to 30-60s after the bad data part. In this case, censoring might have to happen even before the convolution, and still you would make an error, since you assume zero respiration volume and no heartbeats there.

Let me know, if this is detailed enough and what you come up with! We would be very happy to integrate any new suggestions.

All the best, and again my apologies for any inconvenience caused by the lack of this code,
Lars

**Comment by DanielHoffmannAyala** [comment]
Dear Lars,

thank you for your very quick response indeed.
I implemented the idea of flagging unreliable data segments in the design matrix `R` as discussed above, using your code suggestions above and created a function `tapas_physio_censor_unreliable_regressor_parts_retroicor` that would be called within `tapas_physio_main_create_regressors` just after `tapas_physio_create_retroicor_regressors`.

To extend censoring of unreliable data to the respiratory domain, I repeated the same code idea for respiration, all within the same function. I **also** proceeded to censor the variable `mult_sess` in order to ignore also the **Multiplicative Terms** of the Fourier Expansion if **any** of the data at the relevant time points are unreliable.
Is this reasoning sensible?

Unfortunately, this did not result in the censoring I had expected from the pattern of binary values in the variable `physio.ons_secs.c_is_reliable`. Strictly speaking, it there was no censoring at all.
After some digging, I believe I found an incompatibility of `tapas_physio_downsample_phase`. (for details see the attached function)

This is the **code** I came up with (github's upload manager does not support .m files):
[tapas_physio_censor_unreliable_regressor_parts_retroicor.txt](https://github.com/translationalneuromodeling/tapas/files/1731676/tapas_physio_censor_unreliable_regressor_parts_retroicor.txt)
[**tapas_physio_downsample_phase_mod.txt**](https://github.com/translationalneuromodeling/tapas/files/1731724/tapas_physio_downsample_phase_mod.txt)

After modification, censoring worked as expected.
I wonder whether this incompatibility is limited to this specific application we are trying to apply the downsampling function for?

Of course, this implementation is limited to RETROICOR, due to the problems resulting from convolution. (Currently, we do not intend to use HRV and/or RVT)

I would greatly appreciate your feedback on the provided code, to be sure to have implemented the idea correctly.

Thank you again, all the best,
Daniel

**Comment by mrikasper** [comment]
Dear Daniel,

your reasoning with respect to generalization of my code snippet to respiratory and multiplicative effects is totally correct.

I think I now understand the issue you were facing with the wrong timing vectors and why you had to implement a new downsampling function. In fact, there is just a small adjustment necessary in order to use my original suggestion with the standard downsampling function:

```
is_reliable_at_sample_points  = tapas_physio_downsample_phase(ons_secs.t, ...
 is_reliable, sample_points, rsampint);
```
- Instead of ons_secs.spulse, just using **ons_secs.t** as first input to the downsampling will do the trick

 (as is done in `tapas_physio_create_retroicor_regressors` for the respiratory data)
- My copy/past of the cardiac downsampling was indeed misleading, since it relies on the cardiac phase, which is only computed per slice, unlike the filtered respiratory trace (which is more like the outlier vector mask you try to downsample).

In brief, to check this, I implemented the censoring function now also for general use in PhysIO, i.e. -
- with proper interfaces , such as a `physio.model.censor_unreliable_recording_intervals` switch parameter and 
- an informative plots what has been censored (and how the censoring mask was downsampled)

This will become part of the next major release of PhysIO in June (including GUI support), but for direct use now I attached my version of the censoring (and downsampling) function, that you can also just call directly after the call to `tapas_physio_create_retroicor_regressors` in `tapas_physio_main_create_regressors` via:

```
      [ons_secs, cardiac_sess, respire_sess, mult_sess, verbose] = ...
                tapas_physio_censor_unreliable_regressor_parts_retroicor(...
                ons_secs, sqpar, cardiac_sess, respire_sess, mult_sess, verbose);
 ```

Renaming from `.txt` to `.m` will be necessary again for `tapas_physio_censor_unreliable_regressor_parts_retroicor`.

[tapas_physio_censor_unreliable_regressor_parts_retroicor.txt](https://github.com/translationalneuromodeling/tapas/files/1750248/tapas_physio_censor_unreliable_regressor_parts_retroicor.txt)

Maybe you can check whether this function has the expected behavior (and identical to your implementation) and let me know.

Thank you for you input and help,
Lars

**Comment by DanielHoffmannAyala** [comment]
Dear Lars,

the new code you provided behaved as expected and identical to my implementation.
Thank you very much for the efficient dialogue and the comprehensive code suggestions.

Could you please add my affiliation to TU Munich to the credits you so kindly inserted in the code?
I also hope to have some probably helpful ideas on the other open issue soon.

Best,
Daniel

**Comment by mrikasper** [comment]
Dear Daniel,

I know it has been ages, but your suggestion to include your affiliation into the PhysIO file header as well sparked some quite deep questions on ownership and legal consequences for our licensing scheme that had to be harmonized between the whole TAPAS developer team and our institution.

As we now outline in the [CONTRIBUTING Guide](https://github.com/translationalneuromodeling/tapas/blob/master/CONTRIBUTING.md) that is shipped with the most recent version 3.0.0 of TAPAS, we now request external developers to sign (i.e. add their name to) the [Contributor License Agreement](https://github.com/translationalneuromodeling/tapas/blob/master/Contributor-License-Agreement.md) when submitting a code change via merge requests.

The reason, in brief, is to be able to maintain the licensing of TAPAS in the future, without having to collect consent from all contributors again. As a research tool, we would like to keep TAPAS freely available via the GPL 3.0, but there might be other domains in the future, where a different license could be beneficial.

More details are found in the Guide above. What this also entails is a canonical place to give credit to external developers, since they will all be listed within the Contributor-License-Agreement file.

I hope this explanation is still useful and look forward to your future contributions.

All the best,
Lars
